### DISCUSSION TITLE
Where are you using gomponents?

### DISCUSSION THREAD

**Discussion by markuswustenberg** [maragudk/gomponents#163] [General]
I'd love to hear some stories. 😊

Where are you using gomponents? Show off your personal project or describe your work project. 

Where does it shine?

**Comment by mikestefanello** [comment]
Hey @markuswustenberg - Big thanks for creating this wonderful library. It took me a while to come around to this approach over standard templates, but it's really grown on me and I enjoy using it. I recently made a huge change to [Pagoda](https://github.com/mikestefanello/pagoda) to port it from Go templates to Gomponents. I outline how everything is built and structured [here](https://github.com/mikestefanello/pagoda?tab=readme-ov-file#user-interface) (and [here](https://github.com/mikestefanello/pagoda/tree/main/pkg/ui) is the code). Let me know if you have any feedback or ideas to handle things better.

> Where does it shine?

In a lot of places, but for my example, I think the form element _gomponents_ shine the most since it provides inline validation markup without having to do anything extra.

**Reply by markuswustenberg** [reply]
Hey @mikestefanello, thank you for the kind words! I'm very happy that you saw the advantages of gomponents. Reading the part from the readme, you seem to have hit exactly the same pain points I did before writing the library myself. :D

You seem to have a mix of declarative and imperative style of building UI components. An example of the imperative style is constructing a group and appending to it (https://github.com/mikestefanello/pagoda/blob/main/pkg/ui/pages/home.go#L21). I'm definitely not saying it's wrong (!), but I'm curious why you're doing it that way, instead of sticking to declarative (using capital letter `If`)?

**Reply by mikestefanello** [reply]
There's a few reasons:

1) In some cases, like the one you linked to, what you intend to render under certain conditions is not currently available. In this particular case, `r.AuthUser` might be `nil` on any given request, if the user is not logged in. If I use `If()`, the code still executes even if the condition is `false`, and thus would panic in this case. This sort of relates to my next point, but perhaps `If()` could be modified, or another helper provided, that only executes if the condition is `true`.

2) Since the code still executes even if the `If()` condition is `false`, it felt wasteful, at times, to use it, rather than just using a standard `if`. I do use it for more trivial cases, but aside from the `nil` case above, there's a slight waste of performance to execute all of those functions, build slices, allocate strings, etc, to ultimately do nothing. I know for 99.9% of applications, especially anything using a _starter kit_ like Pagoda, this will be completely trivial and negligible, but sometimes I can't help myself. I'm used to dealing with services that handle >1B requests a day, so I'm always thinking about optimizations. In case you missed it in the readme, I also included a mostly unnecessary, though slightly clever, `Node` cache to better handle static content (https://github.com/mikestefanello/pagoda/blob/main/pkg/ui/cache/cache.go). It not only caches a given `Node` in memory, but also fully renders it and converts it to a `Raw` before caching it so it does not have to be re-rendered on every single request. Again, unnecessary and trivial, but it did provide quite a nice improvement in basic benchmarks for the large, static parts of my layouts.

3) The switch to gomponents was [huge](https://github.com/mikestefanello/pagoda/pull/103) and recent and I haven't yet did the usual thing where I continually go over everything I write, looking for ways to clean it up, better organize everything, etc. What's in this `ui` dir was put together pretty fast and I was still mostly new to gomponents at that time.

**Reply by markuswustenberg** [reply]
1. Have a look at [Iff](https://pkg.go.dev/maragu.dev/gomponents#Iff) instead for those use cases. I think it does exactly what you need.
2. Ah, cool! Yeah, makes perfect sense. I usually handle my caching in a different layer, but if you need caching of static partials, that's a neat optimization (when needed).
3. Ah, right, thank you for the context. Makes sense! Code has to bake for a certain amount of time to become really good. :D

And if you have any questions about gomponents going forward, always feel free to reach out, here or elsewhere. 😊

**Reply by mikestefanello** [reply]
Oh, I completely missed `Iff()`! Thanks for the suggestion. I'll definitely start using that.
